### ISSUE TITLE
Different behaviour in pub binder and persistent binder 

### ISSUE THREAD

**Issue by arnim** [gesiscss/orc#23]
As @faflo just reported for [IWAAN/Topic_modelling_C++.ipynb](https://github.com/gesiscss/IWAAN/blob/legacy/Topic_modelling_C%2B%2B.ipynb) cell 10 
```python
topics, _ = display_topics(model_dir, K, voca_pt, tokens_processed, lng, the_page)
```
persistent bHub https://notebooks.gesis.org/services/binder/v2/gh/gesiscss/IWAAN/a8a28ab?filepath=Topic_modelling_C%2B%2B.ipynb throws an error in 

```bash
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-10-ae57cbe4ce69> in <module>
      1 #print output
----> 2 topics, _ = display_topics(model_dir, K, voca_pt, tokens_processed, lng, the_page)

~/BTM/script/topicDisplay.py in display_topics(model_dir, K, voca_pt, tokens_processed, lng, the_page)
     43         wvs = sorted(wvs, key=lambda d:d[1], reverse=True)
     44         #tmps = ' '.join(['%s' % voca[w] for w,v in wvs[:10]])
---> 45         tmps = ' '.join(['%s:%f' % (voca[w],v) for w,v in wvs[:10]])
     46         topics_words.append([str(w) for w,v in wvs[:40]])
     47         rev = []

~/BTM/script/topicDisplay.py in <listcomp>(.0)
     43         wvs = sorted(wvs, key=lambda d:d[1], reverse=True)
     44         #tmps = ' '.join(['%s' % voca[w] for w,v in wvs[:10]])
---> 45         tmps = ' '.join(['%s:%f' % (voca[w],v) for w,v in wvs[:10]])
     46         topics_words.append([str(w) for w,v in wvs[:40]])
     47         rev = []

KeyError: 0
```

while regular bHub https://notebooks.gesis.org/binder/v2/gh/gesiscss/IWAAN/a8a28ab?filepath=Topic_modelling_C%2B%2B.ipynb does not.

**Comment by faflo** [comment]
thanks. the reason for that specific key error is that there is no key 0 in  "voca", while in the regular bHub, there is. One reason I suspect is that something goes wrong in the cell above where the BTM is trained, which takes only 0.6 sec in the persistent Binder (16 sec in regular) and the outputs look off. (BTM/output/model)

**Comment by arnim** [comment]
the difference seems to already occur in the creation of $HOME/BTM/input/input.txt

**Comment by MridulS** [comment]
This is a weird issue indeed.

The issue comes up due to file handling (not sure why it's not the same on both pbhub and bhub). In cell 6 the file `input.txt` is opened but never closed which leads to this behaviour. So I just added `wf.close()` to close the file so it gracefully closes and saves the data to disk. But now Cell 9 gives me an error (coming from the c++ code)
```
ERR: index=13535, size=13535
btm: pvec.h:168: T& Pvec<T>::operator[](int) [with T = double]: Assertion `i < p.size()' failed.
```

To solve this I changed `W = X_max #vocab size` to `W = X_max + 1 #vocab size` [I'm not totally sure what's happening inside the c++ code, not looked into much detail].

Changing these two things seems to make both pbhub and bhub sync up in behaviour.

in cell 6
```
X_max = np.max([np.max(x) for x in X]) #the max token_id
wf = open('BTM/input/input.txt', 'w')
for x in tokens_processed['token_id']:
    print(' '.join(map(str, [str(it) for it in x])), file=wf)
wf.close() # adding this
```

in cell 9
```
W = X_max + 1 #vocab size
```
